spark3/src/main/java/org/apache/iceberg/spark/procedures/BaseProcedure.java

```diff
@@ -56,7 +65,9 @@ protected BaseProcedure(TableCatalog tableCatalog) {
 
     T result = func.apply(icebergTable);
 
-    refreshSparkCache(ident, sparkTable);
+    if (refreshSparkCache) {
+      refreshSparkCache(ident, sparkTable);
+    }
```

Comment:
I think the cache still needs to be refreshed. A commit will refresh the table reference that is cached in Iceberg, pulling in all changes that happened since it was loaded. There could be changes unrelated to the manifest rewrite that get pulled in and would cause the cache to be stale compared with a new query. I think it is best to refresh Spark's cache whenever a procedure modifies or refreshes the table.

Reply:
This raises a good point. I did not consider external modifications and my motivation was to keep data cached if procedures don't modify the underlying data as caching is usually expensive.

Do we think external modifications should invalidate Spark's cache? It seems like a good idea but there is one concern: it does not match how things work right now. If I understand correctly, Spark's cache does not know anything about external modifications and stays valid even if someone modifies the underlying table in Presto or another Spark job. So the only case when the cache is invalidated is when an operation happens in a given session.

I don't feel strongly about this but I want us to think through here.

Reply:
For this PR, I think it is better to invalidate the cache.

The more general question is one for Spark, but I think that the answer is that Spark does not care about external changes. I think that Spark only cares about consistency with its operations. If a table is loaded, then all queries and cached dataframes should give consistent results. If Spark updates, refreshes, or invalidates a table then it should invalidate caches. Otherwise, Spark is simply using the current state as of when the table was loaded, which is a good thing. Spark can't be expected to invalidate a cache every time anything external changes.

Reply:
> If a table is loaded, then all queries and cached dataframes should give consistent results.

I am convinced.

Reply:
Updated.